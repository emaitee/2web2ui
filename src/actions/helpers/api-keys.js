import _ from 'lodash';
import { getGrants } from 'src/selectors/api-keys';

export function formatKeyForRequest(key, getState) {
  const request = { data: {}};

  if (key.subaccount) {
    request.headers = { 'X-MSYS-SUBACCOUNT': key.subaccount.id };
  }

  request.data.label = key.label;

  if (key.grantsRadio === 'all') {
    request.data.grants = _.keys(getGrants(getState()));
  } else {
    request.data.grants = _.keys(_.pickBy(key.grants));
  }

  if (key.validIps) {
    request.data.valid_ips = key.validIps.split(',').map(_.trim);
  }

  return request;
}
