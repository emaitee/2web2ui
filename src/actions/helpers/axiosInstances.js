import axios from 'axios';
import config from 'src/config';

const { apiBase, apiRequestTimeout, zuora: zuoraConfig } = config;

export const sparkpost = axios.create({
  baseURL: apiBase,
  timeout: apiRequestTimeout
});

export const zuora = axios.create({
  baseURL: zuoraConfig.baseUrl,
  timeout: zuoraConfig.timeout
});
