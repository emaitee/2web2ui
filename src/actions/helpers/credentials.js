import { pickBy, trim, keys } from 'lodash';

import { getGrants } from 'selectors/credentials';

export function formatKeyForRequest(key, getState) {
  const request = { data: {}};

  if (key.subaccount) {
    request.headers = { 'X-MSYS-SUBACCOUNT': key.subaccount.id };
  }

  request.data.label = key.label;

  if (key.grantsRadio === 'all') {
    request.data.grants = keys(getGrants(getState()));
  } else {
    request.data.grants = keys(pickBy(key.grants));
  }

  if (key.validIps) {
    request.data.valid_ips = key.validIps.split(',').map(trim);
  }

  return request;
}
