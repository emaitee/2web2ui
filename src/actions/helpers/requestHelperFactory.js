import axios from 'axios';

function defaultOnSuccess({ types, response, dispatch, meta, action }) {
  dispatch({
    type: types.SUCCESS,
    payload: response,
    meta
  });

  return response;
}

function defaultOnFail({ types, err, dispatch, meta, action }) {
  const { message, response = {}} = err;

  dispatch({
    type: types.FAIL,
    payload: { message, response },
    meta
  });
}

export default function requestFactory({
  request = axios,
  onSuccess = defaultOnSuccess,
  onFail = defaultOnFail,
  transformHttpOptions = (opts) => opts
} = {}) {
  return (action) => (dispatch, getState) => {
    const { type = 'NO_TYPE_DEFINED', meta } = action;
    const { url, method = 'get', params, headers, data } = meta;
    const types = {
      PENDING: `${type}_PENDING`,
      SUCCESS: `${type}_SUCCESS`,
      FAIL: `${type}_FAIL`
    };

    // standardize method included in meta from here on out
    meta.method = method.toLowerCase();

    dispatch({
      type: types.PENDING,
      meta
    });

    const httpOptions = {
      method: meta.method,
      url,
      params,
      headers,
      data
    };

    const transformedHttpOptions = transformHttpOptions(httpOptions, getState);

    return request(transformedHttpOptions)
      .then(
        // request succeeded, we only get here if the request returned a 2xx status code
        (response) => onSuccess({ types, response, dispatch, meta, action, getState }),

        // request failed (remember to throw err in your onFail)
        (err) => {
          onFail({ types, err, dispatch, meta, action, getState });
          return { error: err }
        }
      );
  };
}
