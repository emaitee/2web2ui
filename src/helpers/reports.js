import _ from 'lodash';
import qs from 'query-string';
import { getRelativeDates } from 'src/helpers/date';

export function stringifyTypeaheadfilter(filter) {
  const subaccount = filter.type === 'Subaccount' ? `:${filter.id}` : '';
  return `${filter.type}:${filter.value}${subaccount}`;
}

export function dedupeFilters(filters) {
  return _.uniqBy(filters, stringifyTypeaheadfilter);
}

/**
 * Parses search string
 * @param  {string} search - location.search
 * @return {Object}
 *   {
 *     options - options for refresh actions
 *     filters - array of objects ready to be called with reportOptions.addFilter action
 *   }
 */
export function parseSearch(search) {
  if (!search) {
    return { options: {}};
  }

  const { from, to, range, metrics, filters = []} = qs.parse(search);
  const filtersList = (typeof filters === 'string' ? [filters] : filters).map((filter) => {
    const parts = filter.split(':');
    const type = parts.shift();
    let value;
    let id;

    // Subaccount filters include 3 parts
    // 'Subaccount:1234 (ID 554):554' -> { type: 'Subaccount', value: '1234 (ID 554)', id: '554' }
    if (type === 'Subaccount') {
      value = parts[0];
      id = parts[1];
    } else {
      value = parts.join(':');
    }

    return { value, type, id };
  });


  let options = {};

  if (metrics) {
    options.metrics = (typeof metrics === 'string') ? [metrics] : metrics;
  }

  if (from) {
    options.from = new Date(from);
  }

  if (to) {
    options.to = new Date(to);
  }

  if (range) {
    options = { ...options, ...getRelativeDates(range) };
  }

  // filters are used in pages to dispatch updates to Redux store
  return { options, filters: filtersList };
}
