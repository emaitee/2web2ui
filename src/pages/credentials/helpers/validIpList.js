import _ from 'lodash';
import { ipRegex } from 'src/helpers/regex';

export default function validIpList(value) {
  if (!value) {
    return;
  }

  const ips = value.split(',').map(_.trim);
  const invalidIps = ips.filter((ip) => !ipRegex.test(ip));

  if (invalidIps.length) {
    return 'Must be a comma separated list of valid IPs';
  }
}
