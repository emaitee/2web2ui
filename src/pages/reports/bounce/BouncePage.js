import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

import { refreshBounceChartMetrics, refreshBounceTableMetrics } from 'src/actions/bounceReport';
import { addFilter, refreshTypeaheadCache } from 'src/actions/reportFilters';
import { getShareLink, getFilterSearchOptions, parseSearch } from 'src/helpers/reports';
import { showAlert } from 'src/actions/globalAlert';

import { TableCollection, Empty, LongTextContainer } from 'src/components';
import { Percent } from 'src/components/formatters';
import PanelLoading from 'src/components/panelLoading/PanelLoading';
import { Page, Panel } from '@sparkpost/matchbox';
import ShareModal from '../components/ShareModal';
import Filters from '../components/Filters';
import ChartGroup from './components/ChartGroup';

const columns = [{ label: 'Reason', width: '45%' }, 'Domain', 'Category', 'Classification', 'Count (%)'];

export class BouncePage extends Component {
  state = {
    modal: false,
    link: ''
  }

  componentDidMount() {
    this.handleRefresh(this.parseSearch());
    this.props.refreshTypeaheadCache();
  }

  parseSearch() {
    const { options, filters } = parseSearch(this.props.location.search);

    if (filters) {
      filters.forEach(this.props.addFilter);
    }

    return options;
  }

  handleRefresh = (options) => {
    this.props.refreshBounceChartMetrics(options)
      .then(() => this.updateLink())
      .then(() => this.props.refreshBounceTableMetrics(options))
      .catch((err) => {
        this.props.showAlert({ type: 'error', message: 'Unable to refresh bounce report.', details: err.message });
      });
  }

  handleModalToggle = (modal) => {
    this.setState({ modal: !this.state.modal });
  }

  updateLink = () => {
    const { filters, history } = this.props;
    const options = getFilterSearchOptions(filters);
    const { link, search } = getShareLink(options);

    this.setState({ link });
    history.replace({ pathname: '/reports/bounce', search });
  }

  getRowData = (rowData) => {
    const { totalBounces } = this.props;
    const { reason, domain, bounce_category_name, bounce_class_name, count_bounce } = rowData;
    return [
      <LongTextContainer text={reason} />,
      domain,
      bounce_category_name,
      bounce_class_name,
      <span>{count_bounce}(<Percent value={(count_bounce / totalBounces) * 100} />)</span>
    ];
  };

  renderChart() {
    const { chartLoading, aggregates } = this.props;

    if (!chartLoading && !aggregates) {
      return <Empty title='Bounce Rates' message='No bounces to report' />;
    }
    return <ChartGroup />;
  }

  renderCollection() {
    const { tableLoading, reasons } = this.props;

    if (tableLoading) {
      return <PanelLoading />;
    }

    if (!reasons) {
      return <Empty title={'Bounced Messages'} message={'No bounce reasons to report'} />;
    }

    return <TableCollection
      columns={columns}
      rows={reasons}
      getRowData={this.getRowData}
      pagination={true}
    />;
  }

  render() {
    const { modal, link } = this.state;

    return (
      <Page title='Bounce Report'>
        <Filters refresh={this.handleRefresh} onShare={this.handleModalToggle} />
        { this.renderChart() }
        <Panel title='Bounced Messages' className='ReasonsTable'>
          { this.renderCollection() }
        </Panel>
        <ShareModal
          open={modal}
          handleToggle={this.handleModalToggle}
          link={link} />
      </Page>
    );
  }
}

const mapStateToProps = (state) => {
  const chartLoading = state.bounceReport.aggregatesLoading || state.bounceReport.categoriesLoading;
  const tableLoading = chartLoading || state.bounceReport.reasonsLoading;
  const aggregates = state.bounceReport.aggregates;
  return {
    filters: state.reportFilters,
    chartLoading,
    aggregates,
    totalBounces: aggregates ? aggregates.countBounce : 1,
    tableLoading,
    reasons: state.bounceReport.reasons
  };
};

const mapDispatchToProps = {
  addFilter,
  refreshBounceChartMetrics,
  refreshBounceTableMetrics,
  refreshTypeaheadCache,
  showAlert
};
export default withRouter(connect(mapStateToProps, mapDispatchToProps)(BouncePage));
