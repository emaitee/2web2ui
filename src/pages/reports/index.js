import BouncePage from './bounce/BouncePage.js';
import SummaryPage from './summary/SummaryPage.js';
import MessageEventsPage from './messageEvents/MessageEventsPage.js';
import EventPage from './messageEvents/EventPage.js';
import DelayPage from './delays/DelayPage';

export default {
  BouncePage,
  SummaryPage,
  MessageEventsPage,
  EventPage,
  DelayPage
};
