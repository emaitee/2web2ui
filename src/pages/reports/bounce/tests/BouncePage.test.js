import React from 'react';
import { BouncePage } from '../BouncePage';
import { shallow } from 'enzyme';
import * as reportHelpers from 'src/helpers/reports';

jest.mock('src/helpers/reports');

Date.now = jest.fn(() => 1487076708000);

describe('BouncePage: ', () => {

  const props = {
    loading: false,
    aggregates: { countBounce: 1 },
    reasons: [ {
      bounce_category_name: 'Block',
      bounce_class_description: 'The message was blocked by the receiver as coming from a known spam source',
      bounce_class_name: 'Spam Block',
      classification_id: 51,
      count_bounce: 5,
      domain: 'yahoo.com',
      reason: '554 - 5.7.1 Blacklisted by \'twoomail.com\'(twoo.com.multi.surbl.org) Contact the postmaster of this domain for resolution. This attempt has been logged.'
    } ],
    refreshBounceChartMetrics: jest.fn((a) => Promise.resolve()),
    refreshBounceTableMetrics: jest.fn((a) => Promise.resolve()),
    refreshTypeaheadCache: jest.fn((a) => Promise.resolve()),
    addFilter: jest.fn(),
    location: {
      search: {}
    },
    history: {
      replace: jest.fn()
    }
  };

  let wrapper;
  let spyFilterListFromSearch;
  let spyParseSearch;
  let spyGetShare;

  beforeEach(() => {
    spyFilterListFromSearch = reportHelpers.getFilterSearchOptions = jest.fn(() => []);
    spyParseSearch = reportHelpers.parseSearch = jest.fn(() => ({ options: {}}));
    spyGetShare = reportHelpers.getShareLink = jest.fn(() => ({ link: '', search: '' }));
    wrapper = shallow(<BouncePage {...props} />);
  });

  afterEach(() => {
    spyFilterListFromSearch.mockRestore();
    spyParseSearch.mockRestore();
    spyGetShare.mockRestore();
  });

  it('should render', () => {
    const parseSpy = jest.spyOn(wrapper.instance(), 'parseSearch');
    wrapper.instance().componentDidMount();
    expect(props.refreshBounceChartMetrics).toHaveBeenCalled();
    expect(props.refreshBounceTableMetrics).toHaveBeenCalled();
    expect(parseSpy).toHaveBeenCalled();
    expect(props.refreshTypeaheadCache).toHaveBeenCalled();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with no bounces', () => {
    wrapper.setProps({ aggregates: null });
    expect(wrapper).toMatchSnapshot();
  });
});
