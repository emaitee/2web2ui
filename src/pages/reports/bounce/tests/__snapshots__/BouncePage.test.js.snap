// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BouncePage:  should render 1`] = `
<Page
  empty={Object {}}
  title="Bounce Report"
>
  <Connect(Filters)
    onShare={[Function]}
    refresh={[Function]}
  />
  <Connect(ChartGroup) />
  <Panel
    className="ReasonsTable"
    title="Bounced Messages"
  >
    <TableCollection
      columns={
        Array [
          Object {
            "label": "Reason",
            "width": "45%",
          },
          "Domain",
          "Category",
          "Classification",
          "Count (%)",
        ]
      }
      getRowData={[Function]}
      pagination={true}
      rows={
        Array [
          Object {
            "bounce_category_name": "Block",
            "bounce_class_description": "The message was blocked by the receiver as coming from a known spam source",
            "bounce_class_name": "Spam Block",
            "classification_id": 51,
            "count_bounce": 5,
            "domain": "yahoo.com",
            "reason": "554 - 5.7.1 Blacklisted by 'twoomail.com'(twoo.com.multi.surbl.org) Contact the postmaster of this domain for resolution. This attempt has been logged.",
          },
        ]
      }
    />
  </Panel>
  <ShareModal
    handleToggle={[Function]}
    link=""
    open={false}
  />
</Page>
`;

exports[`BouncePage:  should render correctly with no bounces 1`] = `
<Page
  empty={Object {}}
  title="Bounce Report"
>
  <Connect(Filters)
    onShare={[Function]}
    refresh={[Function]}
  />
  <Empty
    message="No bounces to report"
    title="Bounce Rates"
  />
  <Panel
    className="ReasonsTable"
    title="Bounced Messages"
  >
    <TableCollection
      columns={
        Array [
          Object {
            "label": "Reason",
            "width": "45%",
          },
          "Domain",
          "Category",
          "Classification",
          "Count (%)",
        ]
      }
      getRowData={[Function]}
      pagination={true}
      rows={
        Array [
          Object {
            "bounce_category_name": "Block",
            "bounce_class_description": "The message was blocked by the receiver as coming from a known spam source",
            "bounce_class_name": "Spam Block",
            "classification_id": 51,
            "count_bounce": 5,
            "domain": "yahoo.com",
            "reason": "554 - 5.7.1 Blacklisted by 'twoomail.com'(twoo.com.multi.surbl.org) Contact the postmaster of this domain for resolution. This attempt has been logged.",
          },
        ]
      }
    />
  </Panel>
  <ShareModal
    handleToggle={[Function]}
    link=""
    open={false}
  />
</Page>
`;
